File: arrays/reverse_string/reverse_string.c
void reverseString(int* nums, int start, int finish) {
    int front = start;
    int back = finish;
    while (front < back) {
        nums[front] = nums[front] ^ nums[back];
        nums[back] = nums[front] ^ nums[back];
        nums[front] = nums[front] ^ nums[back];
        front += 1;
        back -= 1;
    }
}

void rotate(int* nums, int numsSize, int k) {
    int finish = k % numsSize;
    reverseString(nums, 0, numsSize-1);
    reverseString(nums, 0, finish-1);
    reverseString(nums, finish, numsSize-1);
}

File: arrays/reverse_string/reverse_string.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Void} Do not return anything, modify nums in-place instead.
def reverse_string(nums, start, finish)
    front = start
    back = finish
    while front < back 
        nums[front] = nums[front] ^ nums[back]
        nums[back] = nums[front] ^ nums[back]
        nums[front] = nums[front] ^ nums[back]
        front += 1
        back -= 1
    end
end

def rotate(nums, k)
    length = nums.length
    finish = k % length 
    reverse_string(nums, 0, length-1)
    reverse_string(nums, 0, finish-1)
    reverse_string(nums, finish, length-1)
end
